### ISSUE TITLE
Text missing after component is reattached

### ISSUE THREAD

**Issue by coolersport** [OpenESignForms/vaadin-ckeditor#60]
When a CKEditorTextField is re-attached into a layout, the text isn't sent to client. The initial paintContent() call has set `textIsDirty` to `false`. When it's reattached and repainted, `textIsDirty` flag prevents the text to be sent again.

Suggested fix:
```java
@Override
public void detach() {
    super.detach();
    textIsDirty = true;
}
```

Thanks,

**Comment by OpenESignForms** [comment]
That's interesting, but why does the editor get detached in the first place?  I'm sure we never attempted to concern ourselves with a removed editor to be reused elsewhere vs. just creating a new editor.  Can you explain the scenario better what's going on with the detach?

**Comment by OpenESignForms** [comment]
Did you want to try this JAR (in a ZIP) to see if it resolves your issue.  As we don't know how to create your issue, we could not confirm if this did anything of value.  We did not notice any issues for us.

[vaadinckeditor-7.11.2.jar.zip](https://github.com/OpenESignForms/vaadin-ckeditor/files/1008494/vaadinckeditor-7.11.2.jar.zip)

**Comment by coolersport** [comment]
The editor is added to a TabSheet whose content is lazily loaded. Once the editor tab is visible for the first time, its content/layout will be added (editor is attached). As user switching to another tab, other tabs' content will be removed/detached. Similarly, the same logic applied in other multi-screen layout scenarios.

I've fixed and tested it in the VAADIN6 branch, will commit soon.

**Comment by OpenESignForms** [comment]
We have editors in multiple tabs, and while we don't lazy load, we see no issues when switching tabs.  Is your code removing the editors when the tabsheet is no longer visible too?  If so, wouldn't you recreate the layout when tabsheet is activated again?  Perhaps we don't see it because we do use @PreserveOnRefresh so a refreshed browser will show the last state of the UI rather than a brand new UI.  But that doesn't happen when switching tabs.  Anyway, I guess you don't need our updates as you are still on Vaadin 6 and we're not doing anything there.

As for the "better fix for issue #60" are you saying that you don't need the reset on 'detach' and just those changes there, or you need both?  Again, since we don't see the bug, it's hard for us to add code and then know if it's working or not.

**Comment by OpenESignForms** [comment]
We added a debug statement to the 'detach' method and added 'attach' so we could see them both, and we noted that the detach isn't called when the tab is switched, only when closed (at which time the view should be considered gone and would need to be recreated).  We do have PreserveOnRefresh set for our UI, but I think that only affects a browser reload.  Perhaps this behavior changed in Vaadin 7 from Vaadin 6?  That is, we don't see the detach being called despite having editors in multiple tabs, and we can switch tabs and see everything fine.

We are not against doing the changes, that's for sure, but we're not really sure what we're fixing if we can't see the problem first.

**Comment by coolersport** [comment]
It's because of a different design approach. Our applications contain hundreds of screens in a single application. Pushing all server components to the client to over-fill it with dom nodes led to a sluggish browser performance (tested). Hence only visible screens are left in the dom tree. The server components are useful when users switch back to the previous screen with the most recent state.

For a reusable component like this, it should conform to the component lifecycle than worry about where it fits in. In another word, there is no reason why it should behave any different to a `TextField` component (or any similar) in this scenario.

`PreserveOnRefresh =  true` just makes it behave the same as Vaadin 6.

**Comment by OpenESignForms** [comment]
Thanks for your comments.  

Do you have a simple test UI that can show the issue per your design approach?  

It seems if you were worried about memory usage, you'd create the view when it is made visible and destroy the view when it's no longer visible.  But in your case, it sounds like you have it hanging around somewhere after it's been detached.  We're more than happy to include fixes for scenarios we're not familiar with, but not with code that shows the bug so we can see that the changes fix it.

**Comment by OpenESignForms** [comment]
We never saw how to reproduce this, but we added to the vaadin 7 version on the hopes it resolves something for others too as we've not seen any issue ourselves.  We see no harm (yet) so it should be in the next release.
